hubspot3/test/test_contacts.py

```diff
@@ -1,112 +1,319 @@
+                            "value": "22a26060-c9d7-44b0-9f07-aa40488cfa3a",
+                        },
+                    ],
+                    "vid": 61574,
+                }
+            ],
+            "properties": {},
+        }
+        mock_connection.set_response(200, response_body)
+        resp = contacts_client.create(data)
+        mock_connection.assert_num_requests(1)
+        mock_connection.assert_has_request("POST", "/contacts/v1/contact?", data)
+        assert resp == response_body
+
+    def test_update_by_email(self, contacts_client, mock_connection):
+        email = "mail@email.com"
+        data = {"properties": [{"property": "firstname", "value": "hub"}]}
+        response_body = ""
+        mock_connection.set_response(204, response_body)
+        resp = contacts_client.update_by_email(email, data)
+        mock_connection.assert_num_requests(1)
+        mock_connection.assert_has_request(
+            "POST", "/contacts/v1/contact/email/{}/profile?".format(email), data
+        )
+        assert resp == response_body
+
+    def test_merge(self, contacts_client, mock_connection):
+        primary_id, secondary_id = 10, 20
+        mock_connection.set_response(200, "SUCCESS")
+        resp = contacts_client.merge(primary_id, secondary_id)
+        mock_connection.assert_num_requests(1)
+        mock_connection.assert_has_request(
+            "POST",
+            "/contacts/v1/contact/merge-vids/{}/?".format(primary_id),
+            dict(vidToMerge=secondary_id),
+        )
+        assert resp is None
+
+    @pytest.mark.parametrize(
+        "limit, query_limit, extra_properties",
+        [
+            (0, 100, None),
+            (10, 10, None),
+            (20, 20, None),
+            (150, 100, None),  # keep query limit at max 100
+            (10, 10, ["hs_analytics_last_url", "hs_analytics_revenue"]),
+            (10, 10, "lead_source"),
+        ],
+    )
+    def test_get_all(
+        self, contacts_client, mock_connection, limit, query_limit, extra_properties
+    ):
+        response_body = {
+            "contacts": [
+                {
+                    "addedAt": 1390574181854,
+                    "vid": 204727,
+                    "canonical-vid": 204727,
+                    "merged-vids": [],
+                    "properties": {},
+                }
+            ],
+            "has-more": False,
+            "vid-offset": 204727,
+        }
+        get_batch_return = [dict(id=204727 + i) for i in range(30)]
+        get_batch_mock = Mock(return_value=get_batch_return)
+        contacts_client.get_batch = get_batch_mock
+        mock_connection.set_response(200, response_body)
+        resp = contacts_client.get_all(limit=limit, extra_properties=extra_properties)
+        mock_connection.assert_num_requests(1)
+        mock_connection.assert_has_request(
+            "GET",
+            "/contacts/v1/lists/all/contacts/all?count={}&vidOffset=0".format(
```

Comment:
Das hier verlässt sich ebenfalls auf die Sortierung eines `dict`s - die Parameter können auch in einer anderen Reihenfolge auftauchen. Ich hatte genau für sowas die Utility `assert_query_parameters_in_request` an das Mock-Fixture angehangen, die unabhängig von der Reihenfolge prüft.